File: windows/DotsHarness/src/HarnessPluginKit/HostData.cs
// Copyright (c) 2026 DOTS
// Read-only bridge from the native app into plugins. The app registers named
// snapshot providers; native plugins read
// the current snapshot through `harness.host("<topic>")`. No write path.

namespace HarnessPluginKit;

public interface IHostDataRegistry
{
    /// <summary>Native app: expose <paramref name="topic"/> as a live read-only
    /// snapshot. Re-registering replaces the provider. The delegate runs on every
    /// plugin read, so keep it to already-computed values.</summary>
    void Register(string topic, Func<JsonValue> snapshot);

    void Unregister(string topic);

    /// <summary>Plugin-facing: current snapshot for <paramref name="topic"/>, or null.</summary>
    JsonValue? Snapshot(string topic);

    IReadOnlyList<string> AvailableTopics();
}

public sealed class HostDataRegistry : IHostDataRegistry
{
    private readonly Dictionary<string, Func<JsonValue>> _topics = new(StringComparer.Ordinal);

    public void Register(string topic, Func<JsonValue> snapshot) => _topics[topic] = snapshot;

    public void Unregister(string topic) => _topics.Remove(topic);

    public JsonValue? Snapshot(string topic) =>
        _topics.TryGetValue(topic, out var provider) ? provider() : null;

    public IReadOnlyList<string> AvailableTopics() =>
        _topics.Keys.OrderBy(k => k, StringComparer.Ordinal).ToList();
}

File: macos/DotsHarness/Sources/HarnessPluginKit/HostData.swift
// Copyright (c) 2026 DOTS
// Read-only bridge from the native app into plugins. The app registers named
// snapshot providers; native plugins read
// the current snapshot through `harness.host("<topic>")`. No write path — a
// plugin can look at app state, never mutate it here.

import Foundation

@MainActor
public final class HostDataRegistry {
    private var topics: [String: () -> JSONValue] = [:]

    public init() {}

    /// Native app: expose `topic` as a live read-only snapshot. Re-registering
    /// replaces the provider. The closure runs on every plugin read, so keep it
    /// cheap and return already-computed values.
    public func register(_ topic: String, _ snapshot: @escaping () -> JSONValue) {
        topics[topic] = snapshot
    }

    public func unregister(_ topic: String) {
        topics.removeValue(forKey: topic)
    }

    /// Plugin-facing: current snapshot for `topic`, or nil if unknown.
    public func snapshot(_ topic: String) -> JSONValue? {
        topics[topic]?()
    }

    public func availableTopics() -> [String] {
        topics.keys.sorted()
    }
}
